### DISCUSSION TITLE
Use Broadway to process events in event handlers and process managers

### DISCUSSION THREAD

**Discussion by slashdotdash** [commanded/commanded#576] [Ideas]
[Broadway](https://github.com/plataformatec/broadway) is a concurrent, multi-stage tool for building data ingestion and data processing pipelines. It provides features including back-pressure, automatic acknowledgements, batching, fault tolerance with minimal data loss, graceful shutdown, and partitioning. There are plans to also support rate-limiting, telemetry, and back-off.

The above features means Broadway could be ideally suited to processing events, such as for Commanded event handlers. Currently the event store behaviour module defines how subscriptions to an event stream (or all events) are used. These subscriptions provide acknowledgement for at-least-once event delivery guarantees, but events are only processed serially (one at a time).

### Advantages of using Broadway

Broadway provides both batching and partitioning which could be used to implement Event handler concurrency (#184) and Event handler support for receiving events in batches (#261). It would also support more event sources to be used with Commanded as the input to a Broadway pipeline is a `GenStage` producer. The event store behaviour could be separated between writing/reading events and event sources.

Event handlers could process events from multiple sources in the future when Broadway supports multiple producers. This would help to support multiple concurrent, isolated Commanded apps (#117) where each app has its own event store and an event handler may need to process events from many stores.

### Issues to consider 

- To guarantee event ordering (when desired) events would need to be processed serially, as now, but a partition key could allow concurrency while sill guaranteeing order within a partition (such as per-aggregate). Whether events are processed in order or not, batch sizes, and partition key should all be made user-configurable.
- An event handler can return `:stop` in the `error/3` callback - I'm not sure if this would be possible for a Broadway pipeline and whether this behaviour should be replaced with an explicit nack (not acknowledged), rather than stopping processing?

### Feedback

Do you have any thoughts on this approach?

**Comment by Freyskeyd** [comment]
Hello,

Seems great 😃 

Could we split this proposal into pieces to work on it more easily ?

I'm interested in working on this !

**Comment by slashdotdash** [comment]
@Freyskeyd I've done some exploratory coding with Broadway in the [broadway-show](https://github.com/slashdotdash/broadway-show) repo. The behaviour I was keen to test was being able to stop the pipeline producer on error. This would allow Broadway to fit in with the existing event handler / PM error handling approach in Commanded. Take a look at [`BroadwayShow.Producer`](https://github.com/slashdotdash/broadway-show/blob/187a43cbc789bfcb50681df0cf6fe94b1512feb9/lib/broadway_show/producer.ex) for details.

To use Broadway will require changing the existing `Commanded.EventStore` behaviour so that adapters must provide a compatible `GenStage` producer which Commanded can use within a pipeline it defines (per event handler / PM). 

The steps to start work on this would be:

1. Implement a `GenStage` producer for the in-memory event store adapter included in Commanded.
2. Define a Broadway pipeline for an event handler using the event store producer (in-memory) where the `handle_message` or `handle_batch` callback functions delegate to the event handler module (as is currently done in the existing `handle_event/3` function).
3. Replace subscribing to events (using `EventStore.subscribe_to`) in `Commanded.Event.Handler` with having it start the Broadway pipeline. It may even be possible to change `Commanded.Event.Handler` from a `GenServer` to a `Broadway` pipeline. This would mean changing the `start_link/1` function to start the pipeline.

Initially we should try to get the above working with a pipeline configured to use a single stage and batch size of 1. Afterwards we can look at allowing the event handler to configure concurrency and/or batch size.

**Comment by jsmestad** [comment]
@slashdotdash is this still something you are going to pursue?

**Comment by julienmarie** [comment]
That would be quite amazing
